Add GDPR compliance requirement check

mkevins opened the pull request:
Fixes https://github.com/wordpress-mobile/WordPress-Android/issues/18476

### Description

This is an alternative implementation of the work started in this PR: https://github.com/wordpress-mobile/WordPress-Android/pull/18592. This implementation uses the same underlying source as WooCommerce (i.e. country code is fetched, rather than obtained from the OS).

To test:
Run unit tests

## Regression Notes
1. Potential unintended areas of impact
None

2. What I did to test those areas of impact (or what existing automated tests I relied on)
This should not affect any existing UI (since it implements a new, and not-yet-used data layer)

3. What automated tests I added (or what prevented me from doing so)
This PR adds some unit tests to cover scenarios for when we should show the privacy banner.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

mkevins committed:
Add event bus coroutines utility

mkevins committed:
Add deletable preference for banned already shown state

By making this a deletable preference, the banner may be shown again
after a user logs out and logs back in. This is a trade-off we accept to
ensure that we don't omit the banner for a different user, since this is
saved as an app-level preference.

mkevins committed:
Add geo endpoint support

This also provides the logic to check for GDPR compliance requirements.

mkevins committed:
Add logic to check whether to ask for privacy consent

This relies on several factors, including country code and logged in
state. Various scenarios are covered in unit tests.

mkevins committed:
Add privacy settings repository

This handles remotely persisting the privacy settings to the account
settings endpoint.

ovitrif commented on WordPress/src/main/java/org/wordpress/android/util/EventBusCoroutines.kt:
Finally, we can make the term `usecase` great again and not pollute our code with weird constructs making it look as if a use case means a mapper from event-bus to coroutines.

Not sure if this is from woo-android or your own creation, but it is a long-awaited masterpiece we actually needed for years. Glad to see it making it to our code, next thing it can enable is to replace all weird `xUseCase` classes with implementations using these utils, and start using the term for what it really is: a class/code where business logic is done, not an adapter (ie. coroutines from eventbus).

mkevins replied:
I couldn't help but have the same thought when discovering this code. 😄  The work seems to have started here: https://github.com/woocommerce/woocommerce-android/pull/7587, and has evolved on the Woo side a bit since then. H/t to Hicham Boushaba for implementing this 🥇

ovitrif commented on WordPress/src/test/java/org/wordpress/android/ui/prefs/privacy/banner/domain/ShouldAskPrivacyConsentTest.kt:
This is a really great piece of code, looks lovely, reads awesomely!

More of this, I would love to see :Yoda:

ovitrif reviewed (approved):
Great work @mkevins!

All for shipping it 👍🏻

ovitrif committed:
Fix: detekt issue

mkevins committed:
Merge branch 'trunk' into issue/18476-gdpr-compliance-requirement-check

mkevins committed:
Update FluxC version reference

mkevins committed:
Enable error type inferring in stubs

This is enabled to address compile error:
`ERROR : error.NonExistentClass`. It is expected to lead to better error
messages, but in this case it somehow _fixes_ the compile error.

ovitrif reviewed (approved):
Everything passes for this one, great work as earlier mentioned!
Personally I'm happy with this code 🙇🏻

mkevins commented:
Thank you for reviewing and testing this Ovi, and also, thank you for your initial work on this!

ovitrif commented:
> Thank you for reviewing and testing this Ovi, and also, thank you for your initial work on this!

Thank YOU Matthew for improving and getting this ready!! 🙇🏻
